apps/rush-lib/src/logic/PackageChangeAnalyzer.ts

```diff
@@ -26,14 +28,46 @@ export class PackageChangeAnalyzer {
     this._data = this._getData();
   }
 
-  public getPackageDepsHash(projectName: string): IPackageDeps | undefined {
+  public getPackageDeps(projectName: string): IPackageDeps | undefined {
     if (!this._data) {
       this._data = this._getData();
     }
 
     return this._data.get(projectName);
   }
 
+  /**
+   * The project state hash is calculated in the following way:
+   * - Project dependencies are collected (see PackageChangeAnalyzer.getPackageDeps)
+   *   - If project dependencies cannot be collected (i.e. - if Git isn't available),
+   *     this function returns `undefined`
+   * - The (path separator normalized) repo-root-relative dependencies' file paths are sorted
+   * - A SHA1 hash is created and each (sorted) file path is fed into the hash and then its
+   *   Git SHA is fed into the hash
+   * - A hex digest of the hash is returned
+   */
+  public getProjectStateHash(projectName: string): string | undefined {
+    let projectState: string | undefined = this._projectStateCache.get(projectName);
+    if (!projectState) {
+      const packageDeps: IPackageDeps | undefined = this.getPackageDeps(projectName);
+      if (!packageDeps) {
+        return undefined;
+      } else {
+        const sortedPackageDepsFiles: string[] = Object.keys(packageDeps.files).sort();
+        const hash: crypto.Hash = crypto.createHash('sha1');
+        for (const packageDepsFile of sortedPackageDepsFiles) {
+          hash.update(packageDepsFile);
+          hash.update(packageDeps.files[packageDepsFile]);
```

Comment:
We probably should put a delimiter between the hashed strings.  The delimiter can be any character, just something that cannot possibly occur in any of the hashed strings.  For example:
```
hash.update("|" + packageDepsFile);
```
or
```
hash.update(packageDepsFile + "\n");
```

If we don't do this, then the hash is susceptible to collisions involving string concatenation.  For example, these cases will all produce identical hashes:

Case 1
```
hash.update('ab');
hash.update('c');
```

Case 2
```
hash.update('a');
hash.update('bc');
```

Case 3
```
hash.update('abc');
hash.update('');
```

Such a collision is "unlikely," but not in the statistical sense of a hash algorithm. In practice, it is actually rather likely that new inputs will be hooked up to this code that we didn't think about.

Reply:
(Might also be worth checking the other `crypto.createHash()` calls in our code base for this issue.)

Reply:
Good call. Updated.